language/tools/move-cli/src/main.rs

```diff
@@ -0,0 +1,400 @@
+    let move_data = maybe_create_dir(&args.move_data)?;
+
+    // allow user modules in move_lib directory to compile against stdlib and move_data modules, but
+    // not Libra framework modules
+    // TODO: prevent cyclic dep?
+    let mut user_lib_files: Vec<String> = fs::read_dir(move_lib)?
+        .map(|f| f.map_or("".to_string(), |f| f.path().to_string_lossy().into_owned()))
+        .filter(|p| p.ends_with(".move"))
+        .collect();
+
+    if !user_lib_files.is_empty() {
+        println!("Compiling {:?} user module(s)", user_lib_files.len());
+    }
+    if let Some(f) = script_file.as_ref() {
+        user_lib_files.push(f.to_string())
+    }
+
+    // allow script to compile against stdlib and user move_lib modules, but not Libra framework
+    // modules
+    let deps = if args.no_use_stdlib {
+        vec![]
+    } else {
+        stdlib::stdlib_source_files()
+    };
+    let code_address = Some(Address::MOVE_LIB);
+    let (_, compilation_units) = move_lang::move_compile(&user_lib_files, &deps, code_address)?;
+
+    let mut stdlib_modules: Vec<CompiledModule> = if args.no_use_stdlib {
+        vec![]
+    } else {
+        stdlib::stdlib_bytecode_files()
+            .iter()
+            .map(|f| {
+                let bytecode_bytes = fs::read(&f)?;
+                CompiledModule::deserialize(&bytecode_bytes)
+                    .map_err(|e| anyhow!("Error deserializing module: {:?}", e))
+            })
+            .collect::<Result<Vec<CompiledModule>>>()?
+    };
+
+    let mut script_opt = None;
+    for c in compilation_units {
+        match c {
+            CompiledUnit::Script { script, .. } => {
+                if script_opt.is_some() {
+                    bail!("Error: found script in move_lib")
+                }
+                script_opt = Some(script)
+            }
+            CompiledUnit::Module { module, .. } => stdlib_modules.push(module),
+        }
+    }
+    Ok((
+        OnDiskStateView::create(move_data.to_path_buf(), &stdlib_modules)?,
+        script_opt,
+    ))
+}
+
+fn run(args: &Move, script_file: &str, signers: &[String], commit: bool) -> Result<()> {
+    let (state, script_opt) = compile(args, &Some(script_file.to_string()))?;
+    let mut script_bytes = vec![];
+    if let Some(script) = script_opt {
+        script.serialize(&mut script_bytes)?
+    } else {
+        bail!(
+            "Script file {:?} contains module instead of script",
+            script_file
+        )
+    }
+
+    let vm = MoveVM::new();
+    // TODO: use nonzero schedule and pick reasonable max default gas price
+    let cost_schedule = gas_schedule::zero_cost_schedule();
+    let mut cost_strategy = gas_schedule::CostStrategy::system(&cost_schedule, GasUnits::new(0));
+    let mut signers = signers
+        .iter()
+        .map(|s| AccountAddress::from_hex_literal(&s))
+        .collect::<Result<Vec<AccountAddress>>>()?;
+    signers.reverse(); // TODO: figure out why VM reads this right-to-left
```

Comment:
CC @vgao1996 in case the answer to this is known

Reply:
```
// VMRuntime::execute_script
            senders.into_iter().for_each(|addr| {
                args.insert(0, Value::transaction_argument_signer_reference(addr))
            });
```
Seems like a bug in the current implementation

Reply:
Hmm, wondering how the `AdminScript` tests work given this... @tzakian might know?

Reply:
In the Libra VM when we're populating the signers for the transaction they're populated "from left to right" so
```
senders == vec![execute_as_signer, libra_root_signer]
```

And is going through this and pushing those on at the front, so you'll get:

```
vec![libra_root_signer, execute_as_signer, <rest of args>]
```

which is what we expect (LR as the first signer). 

We could build up the signers a different way if we wanted to, but I'm not sure f that buys us anything?

Reply:
I think the thing that would be nice to buy us is that the caller of `execute_script` can pass `vec![signer1, signer2]` for a script like `fun main(signer1: &signer, signer2: &signer)` instead of `vec![signer2, signer1]`. "Pass the signers in the reverse order that the script expects them" seems like a weird API, particularly given that we ask the for the args and type args in non-reverse order...

I would propose to do this by swapping 
```
senders.into_iter().for_each(|addr| {
  args.insert(0, Value::transaction_argument_signer_reference(addr))
});
``` 

for something like

```
let mut signers_and_args = senders.into_iter().map(|addr| Value::transaction_argument_signer_reference(addr)).collect();
signers_and_args.append(&mut args);
...
```

Reply:
I don't think the Libra VM should be doing signers *at all* which might be part of the problem

Reply:
Far point! I think the non-reverse order argument makes a good case for why this should be changed. I would be more than happy to see the change go in, so feel free to change in this PR (it shouldn't be a large change) or I can put up a separate PR if that's preferable.

Reply:
I am too chicken to change it in this PR, but am happy to put up a separate one.